base/abstractarray.jl

```diff
@@ -116,41 +116,53 @@ linearindexing{T<:Range}(::Type{T}) = LinearFast()
 *(::LinearFast, ::LinearSlow) = LinearSlow()
 *(::LinearSlow, ::LinearSlow) = LinearSlow()
 
+# The real @inline macro is not available this early in the bootstrap, so this
+# internal macro splices the meta Expr directly into the function body.
+macro _inline_meta()
+    Expr(:meta, :inline)
+end
+macro _noinline_meta()
+    Expr(:meta, :noinline)
+end
+
 ## Bounds checking ##
-checkbounds(sz::Int, ::Colon) = nothing
-checkbounds(sz::Int, i::Int) = 1 <= i <= sz || throw(BoundsError())
-checkbounds(sz::Int, i::Real) = checkbounds(sz, to_index(i))
-checkbounds(sz::Int, I::AbstractVector{Bool}) = length(I) == sz || throw(BoundsError())
-checkbounds(sz::Int, r::Range{Int}) = isempty(r) || (minimum(r) >= 1 && maximum(r) <= sz) || throw(BoundsError())
-checkbounds{T<:Real}(sz::Int, r::Range{T}) = checkbounds(sz, to_index(r))
-
-function checkbounds{T <: Real}(sz::Int, I::AbstractArray{T})
+_checkbounds(sz::Int, i::Int) = 1 <= i <= sz
+_checkbounds(sz::Int, i::Real) = (@_inline_meta; _checkbounds(sz, to_index(i)))
+_checkbounds(sz::Int, I::AbstractVector{Bool}) = length(I) == sz
+_checkbounds(sz::Int, I::AbstractArray{Bool}) = length(I) == sz # setindex! allows this
+_checkbounds(sz::Int, r::Range{Int}) = (@_inline_meta; isempty(r) || (minimum(r) >= 1 && maximum(r) <= sz))
+_checkbounds{T<:Real}(sz::Int, r::Range{T}) = (@_inline_meta; _checkbounds(sz, to_index(r)))
+_checkbounds(sz::Int, ::Colon) = true
+function _checkbounds{T <: Real}(sz::Int, I::AbstractArray{T})
+    @_inline_meta
```

Comment:
Yes, that is a very good question, and it'd probably be good to review all the little performance tweaks I've been making to make sure they're necessary.  This one, however, makes a huge difference.  It's on the order of 10x more costly to call this function than it is to inline it.  These bounds checks are effectively "inlined" into the current n-ary `arrayref` intrinsics, so I needed to do all sorts of these inlining tricks to match it.  These sorts of optimizations are *so* finicky and tricky (and time consuming!) to test.

I've also played with changing the branching behavior (e.g., using `&` instead of `&&`, etc) and can get really nice-looking (and short!) LLVM code that ends up performing worse for some strange cache/pipeline/who-knows-what reason.